lib/checkcondition.cpp

```diff
@@ -1188,12 +1187,13 @@ void CheckCondition::alwaysTrueFalse()
                 continue;
             if (Token::Match(tok, "! %num%|%bool%"))
                 continue;
+            if (Token::Match(tok, "&&|%oror%") && !Token::Match(tok->astOperand1(), "%char%") && !Token::Match(tok->astOperand2(), "%char%"))
+                continue;
 
             const bool constIfWhileExpression =
-                tok->astParent()
-                && Token::Match(tok->astParent()->astOperand1(), "if|while")
-                && !tok->isBoolean();
-            const bool constValExpr = Token::Match(tok, "%num%|%char%") && Token::Match(tok->astParent(),"%oror%|&&|?"); // just one number or char in boolean expression
+                tok->astTop() && Token::Match(tok->astTop()->astOperand1(), "if|while") &&
```

Comment:
In the first check it's enough to check `tok->astParent()`. If there is a `tok->astParent()` then there is a `tok->astTop()` also. The `tok->astParent()` is faster.

Reply:
I meant:

    tok->astParent() && Token::Match(tok->astTop()->astOperand1(), "if|while") && ...